src/main/flight/autopilot.c

```diff
@@ -91,17 +206,182 @@ void altitudeControl(float targetAltitudeCm, float taskIntervalS, float vertical
+    // function is intended for only small changes in position
+    // for example, where the step distance change reflects an intended velocity, determined by a client function
+    // if we had a 'target_ground_speed' value, like in gps_rescue, we can make a function that starts and stops smoothly and targets that velocity
+}
+
+void updateLocation(earthFrame_t *efAxis, axisEF_t loopAxis)
+{
+    if (loopAxis == EW) {
+        currentTargetLocation.lon = gpsSol.llh.lon; // update East-West / / longitude position
+    } else {
+        currentTargetLocation.lat = gpsSol.llh.lat; // update North-South / latitude position
+    }
+    efAxis->previousDistance = 0.0f; // and reset the previous distance
+}
+
+bool positionControl(void) 
+{
+    static uint16_t gpsStamp = 0;
+    if (gpsHasNewData(&gpsStamp)) {
+        ap.gpsDataIntervalS = getGpsDataIntervalSeconds(); // interval for current GPS data value 0.05 - 2.5s
+        ap.gpsDataFreqHz = getGpsDataFrequencyHz();
+
+        // first get NS and EW distances from current location (gpsSol.llh) to target location
+        vector2_t gpsDistance;
+        GPS_latLongVectors(&gpsSol.llh, &currentTargetLocation, &gpsDistance.x, &gpsDistance.y); // X is EW, Y is NS
+        ap.efAxis[EW].distance = gpsDistance.x;
+        ap.efAxis[NS].distance = gpsDistance.y;
+
+        const float distanceCm = vector2Norm(&gpsDistance);
+
+        const float leak = 1.0f - 0.4f * ap.gpsDataIntervalS;
+        // leak iTerm while sticks are centered, 2s time constant approximately
+        const float lpfGain = pt1FilterGain(ap.pt1Cutoff, ap.gpsDataIntervalS);
+
+        // ** Sanity check **
+        // primarily to detect flyaway from no Mag or badly oriented Mag
+        // must accept some overshoot at the start, especially if entering at high speed
+        if (distanceCm > ap.sanityCheckDistance) {
+            return false;
+        }
+
+        static float prevPidDASquared = 0.0f; // if we limit DA on true vector length
+        const float maxDAAngle = 35.0f; // limit in degrees; arbitrary angle
+        const float sqMaxDAAngle = sq(maxDAAngle);
+
+        for (axisEF_t loopAxis = EW; loopAxis <= NS; loopAxis++) {
+            earthFrame_t *efAxis = &ap.efAxis[loopAxis];
+            // separate PID controllers for latitude (NorthSouth or NS) and longitude (EastWest or EW)
+
+            // ** P **
+            const float pidP = efAxis->distance * positionPidCoeffs.Kp;
+
+            // ** I **
+            efAxis->integral += efAxis->isStopping ? 0.0f : efAxis->distance * ap.gpsDataIntervalS;
+            // only add to iTerm while in hold phase
+            const float pidI = efAxis->integral * positionPidCoeffs.Ki;
+
+            // ** D ** //
+            // Velocity derived from GPS position works better than module supplied GPS Speed and Heading information
+
+            float velocity = (efAxis->distance - efAxis->previousDistance) * ap.gpsDataFreqHz; // cm/s, minimum step 11.1 cm/s
```

Comment:
The intent is to explain what happens when this value for velocity is logged, because in the log we see steps in multiples of approximately 11 cm/s, and it isn't obvious why they would be of that size without knowing the resolution of the velocity element.

Reply:
(It will be smaller when flying east on high latitudes)